motionplan/check.go

```diff
@@ -0,0 +1,440 @@
+	// something expressly forbidding it.
+	validateRelPiF := func(pif *referenceframe.PoseInFrame) error {
+		observingFrame := fs.Frame(pif.Parent())
+		// Ensure the frame of the pose-in-frame is in the frame system
+		if observingFrame == nil {
+			sfPlanner.logger.Errorf(
+				"pose of %s was given in frame of %s, but no frame with that name was found in the frame system",
+				checkFrame.Name(),
+				pif.Parent(),
+			)
+			return nil
+		}
+		// Ensure nothing between the PiF's frame and World is the relative frame
+		observingParentage, err := fs.TracebackFrame(observingFrame)
+		if err != nil {
+			return err
+		}
+		for _, parent := range observingParentage {
+			if parent.Name() == checkFrame.Name() {
+				return fmt.Errorf(
+					"pose of %s was given in frame of %s, but current pose of checked frame must not be observed by self or child",
+					checkFrame.Name(),
+					pif.Parent(),
+				)
+			}
+		}
+		return nil
+	}
+
+	toWorld := func(pif *referenceframe.PoseInFrame, inputs map[string][]referenceframe.Input) (*referenceframe.PoseInFrame, error) {
+		// Check our current position is valid
+		err := validateRelPiF(pif)
+		if err != nil {
+			return nil, err
+		}
+		transformable, err := fs.Transform(inputs, pif, referenceframe.World)
+		if err != nil {
+			return nil, err
+		}
+		poseInWorld, ok := transformable.(*referenceframe.PoseInFrame)
+		if !ok {
+			// Should never happen
+			return nil, errors.New("could not convert transformable to a PoseInFrame")
+		}
+		return poseInWorld, nil
+	}
+
+	// Check out path pose is valid
+	stepEndPiF, ok := plan.Path()[wayPointIdx][checkFrame.Name()]
+	if !ok {
+		return errors.New("check frame given not in plan Path map")
+	}
+	expectedArcEndInWorld, err := toWorld(stepEndPiF, plan.Trajectory()[wayPointIdx])
+	if err != nil {
+		return err
+	}
+
+	currentPoseInWorld, err := toWorld(currentPoseIF, currentInputs)
+	if err != nil {
+		return err
+	}
+
+	arcInputs, ok := plan.Trajectory()[wayPointIdx][checkFrame.Name()]
+	if !ok {
+		return errors.New("given checkFrame had no inputs in trajectory map at current index")
+	}
+	fullArcPose, err := checkFrame.Transform(arcInputs)
+	if err != nil {
+		return err
+	}
+
+	// Relative current inputs will give us the arc the base has executed. Calculating that transform and subtracting it from the
+	// arc end position (that is, the same-index node in plan.Path()) gives us our expected location.
+	frameCurrentInputs, ok := currentInputs[checkFrame.Name()]
+	if !ok {
+		return errors.New("given checkFrame had no inputs in CurrentInputs map")
+	}
+	poseThroughArc, err := checkFrame.Transform(frameCurrentInputs)
+	if err != nil {
+		return err
+	}
+	remainingArcPose := spatialmath.PoseBetween(poseThroughArc, fullArcPose)
+	expectedCurrentPose := spatialmath.PoseBetweenInverse(remainingArcPose, expectedArcEndInWorld.Pose())
+	errorState := spatialmath.PoseBetween(expectedCurrentPose, currentPoseInWorld.Pose())
```

Comment:
I'm not sure how I feel about the errorState being calculated here. 
If we have an errorState method on the PTG driver why not use that value instead of re-doing the math here?

Reply:
Timing issues and multiple sources of truth.

What happens when the error state returned does not quite match with what we calculate here?

Reply:
good point, I will need to think about how this can happen and what the best approach would be.

Reply:
My preference is to actually deprecate ErrorState and migrate everything to ExecutionState.

Reply:
> My preference is to actually deprecate ErrorState and migrate everything to use ExecutionState.

I am sold on this idea.
May you please create a ticket so that this is tracked?

Reply:
RSDK-7555